### PR TITLE
[server] added limit offset validation checks

### PR DESCRIPTION
**Pull request by shounakmk219**
#### Issue(s)

https://cortexdata.atlassian.net/browse/TE-1487

#### Description

Earlier when an invalid limit or offset value/combination was passed to the get all endpoints it used to respond with empty list. Now the endpoints will return an error list.

Example
**API call** : `/api/anomalies?isChild=false&offset=5`
**Response code**: `400`
**Response**:
```
{
  "list": [
    {
      "code":"ERR_OFFSET_WITHOUT_LIMIT",
      "msg":"'offset' value provided without 'limit' value."
    }
  ]
}
```

### CODE REVIEW

File: thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/dao/GenericPojoDao.java [hunk 1]
```diff
@@ -68,6 +72,22 @@ public GenericPojoDao(final DatabaseService databaseService,
         "Entity Metadata is inconsistent!");
   }
 
+  private static void validateLimitOffset(final DaoFilter filter) {
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/main/java/ai/startree/thirdeye/datalayer/dao/GenericPojoDao.java:75]: I think you can just call this `validate`

Please use `Preconditions.checkArgument()`, not `ThirdEyeException`. The latter is primarily a framework for communicating with UI.

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java [hunk 2]
```diff
@@ -148,19 +152,15 @@ public void filterWithOffsetWithoutLimitTest() {
     final DaoFilter filter = new DaoFilter()
         .setOffset(offset)
         .setBeanClass(AnomalyDTO.class);
-    final List<AnomalyDTO> anomalies = dao.filter(filter);
-    assertThat(anomalies).isNotNull();
-    assertThat(anomalies).isEmpty();
+    assertThatThrownBy(() -> dao.filter(filter)).hasMessage(ERR_OFFSET_WITHOUT_LIMIT.getMessage());
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:155]: nice! just update this to `IllegalArgumentException` as per above comment. We'll use standard exceptions in the backend and use ThirdEyeException to communicate that automatically to the UI.

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java [hunk 3]
```diff
@@ -148,19 +152,15 @@ public void filterWithOffsetWithoutLimitTest() {
...
 
   @Test
   public void testNegativeLimitValue() {
     final DaoFilter filter = new DaoFilter()
         .setLimit(-5L)
         .setBeanClass(AnomalyDTO.class);
-    final List<AnomalyDTO> anomalies = dao.filter(filter);
-    assertThat(anomalies).isNotNull();
-    assertThat(anomalies).isEmpty();
+    assertThatThrownBy(() -> dao.filter(filter)).hasMessage(ERR_NEGATIVE_LIMIT_VALUE.getMessage());
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:163]: same comment as above

File: thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java [hunk 4]
```diff
@@ -169,9 +169,7 @@ public void testNegativeOffsetValue() {
         .setLimit((long) TOTAL_ANOMALIES)
         .setOffset(-5L)
         .setBeanClass(AnomalyDTO.class);
-    final List<AnomalyDTO> anomalies = dao.filter(filter);
-    assertThat(anomalies).isNotNull();
-    assertThat(anomalies).isEmpty();
+    assertThatThrownBy(() -> dao.filter(filter)).hasMessage(ERR_NEGATIVE_OFFSET_VALUE.getMessage());
```
Comment (suvodeep-pyne) [inline thirdeye-persistence/src/test/java/ai/startree/thirdeye/datalayer/dao/TestGenericPojoDao.java:172]: same comment as above

File: thirdeye-server/src/main/java/ai/startree/thirdeye/DaoFilterBuilder.java [hunk 5]
```diff
@@ -81,10 +86,17 @@ public DaoFilter buildFilter(final MultivaluedMap<String, String> queryParameter
     final DaoFilter daoFilter = new DaoFilter();
     optional(queryParameters.getFirst("limit"))
         .map(Long::valueOf)
-        .ifPresent(daoFilter::setLimit);
+        .ifPresent( limit -> {
+          ensure(limit >= 0, ERR_NEGATIVE_LIMIT_VALUE);
```
Comment (suvodeep-pyne) [inline thirdeye-server/src/main/java/ai/startree/thirdeye/DaoFilterBuilder.java:90]: nice!

File: thirdeye-server/src/main/java/ai/startree/thirdeye/DaoFilterBuilder.java [hunk 6]
```diff
@@ -81,10 +86,17 @@ public DaoFilter buildFilter(final MultivaluedMap<String, String> queryParameter
...
+          daoFilter.setLimit(limit);
+        });
     optional(queryParameters.getFirst("offset"))
         .map(Long::valueOf)
-        .ifPresent(daoFilter::setOffset);
+        .ifPresent(offset -> {
+          ensureExists(daoFilter.getLimit(),ERR_OFFSET_WITHOUT_LIMIT);
```
Comment (suvodeep-pyne) [inline thirdeye-server/src/main/java/ai/startree/thirdeye/DaoFilterBuilder.java:96]: nice! very specific msg helps guide the user to know and fix the problem.

**Review by suvodeep-pyne — APPROVED** [review]
mostly LGTM. Please see comments.

File: thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/AnomalyPaginationTest.java [hunk 7]
```diff
@@ -121,25 +126,28 @@ public void testResponseWithLimitAndOffsetFilters() {
   @Test
   public void testNegativeLimitValue() {
     final Response response = request("api/anomalies?limit=-10").get();
-    final List<AnomalyApi> results = response.readEntity(ANOMALY_LIST_TYPE);
-    assertThat(response.getStatus()).isEqualTo(200);
-    assertThat(results).isEmpty();
+    final StatusListApi results = response.readEntity(ERROR_LIST_TYPE);
```
[the remark this answers is not in the document]
Comment (shounakmk219) [reply thirdeye-integration-tests/src/test/java/ai/startree/thirdeye/AnomalyPaginationTest.java:129]: I have added the unit tests now. Let's keep the integration tests too.

**Pull request merged**